### ISSUE TITLE
Improved Loader Interface

### ISSUE THREAD

**Issue by paul-schaaf** [project-serum/anchor#997]
Currently, there is a Loader type (and an accountloader?) used for zero-copy. This type has a different API than other accounts. In particular, you have to call methods like `load` or `load_mut` instead of just getting a reference to it.

Ideally, the API of the Loader type should mirror the interface of the other accounts. This can probably be done by implementing deref and derefmut for the Loader type in a way that calls these methods internally, so for the user the interface is like all other accounts.

**Comment by archseer** [comment]
A load isn't exactly cost free though: https://github.com/project-serum/anchor/blob/54f24e55960a5452e2bb8a33086aa85e8e9f1bc6/lang/src/loader.rs#L73-L83

It's better to explicitly call `load()` to control when this happens and that it only happens once during execution. If this was hidden inside deref it's very easy to accidentally do this per use.

**Comment by paul-schaaf** [comment]
I agree that moving the logic into deref will pose problems regarding performance if called multiple times.

That said, I think that mostly advanced users will use Loader and those users will look into the documentation before and just dereference to a local variable once and use that one from there.

But even if not and it's used by newcomers, I assume (pls do correct me here) they will only call Loader.load() once in most cases. In case they do call it more than once we could have a flag in the loader struct that prevents multiple`load` calls at runtime when calling `load` (the first time deref is called that flag would be set to true and no more deref would be possible). For those use cases where it actually makes sense to call `load` twice and you know what you're doing, we could have a command/constraint that stops the load check mentioned above from erroring.

My feeling is that with this API which is like a normal account we would satisfy users most of the time and leave advanced users the option to customise.

What do you think @archseer

**Comment by archseer** [comment]
I think it falls outside of intended use for Deref:

> Implementing Deref for smart pointers makes accessing the data behind them convenient, which is why they implement Deref. On the other hand, the rules regarding Deref and DerefMut were designed specifically to accommodate smart pointers. Because of this, Deref **should only be implemented for smart pointers** to avoid confusion.
> 
> For similar reasons, **this trait should never fail.** Failure during dereferencing can be extremely confusing when Deref is invoked implicitly.

https://doc.rust-lang.org/std/ops/trait.Deref.html
